module/vision/BallDetector/src/BallDetector.cpp

```diff
@@ -117,126 +118,155 @@ namespace module::vision {
 
                 log<NUClear::DEBUG>(fmt::format("Found {} clusters below green horizon", clusters.size()));
 
+                // Create the Balls message, which will contain a Ball for every cluster that is a valid ball
                 auto balls = std::make_unique<Balls>();
 
-                // We still emit if there are no balls otherwise other modules are not aware that we can no longer see a
-                // ball, and they will keep using the last message sent. See head behaviour, where we use 'with' balls
-                // and react based on that last balls message.
+                // Balls is still emitted even if there are no balls, to indicate to other modules that no balls are
+                // currently visible
                 if (clusters.empty()) {
                     log<NUClear::DEBUG>("Found no balls.");
                     emit(std::move(balls));
                     return;
                 }
 
+                // Reserve the memory prematurely for efficiency
                 balls->balls.reserve(clusters.size());
 
-                balls->id        = horizon.id;
-                balls->timestamp = horizon.timestamp;
-                balls->Hcw       = horizon.Hcw;
+                balls->id        = horizon.id;         // subsumption id
+                balls->timestamp = horizon.timestamp;  // time when the image was taken
+                balls->Hcw       = horizon.Hcw;        // world to camera transform at the time the image was taken
 
+                // Check each cluster to see if it's a valid ball
                 for (auto& cluster : clusters) {
                     Ball b;
 
                     // Average the cluster to get the cones axis
                     Eigen::Vector3f axis = Eigen::Vector3f::Zero();
+                    // Add up all the unit vectors of each point (camera to point in world space) in the cluster
                     for (const auto& idx : cluster) {
                         axis += rays.col(idx);
                     }
-                    axis /= cluster.size();
+                    axis /= cluster.size();  // can this line be skipped if it's being normalised in the next line?
                     axis.normalize();
 
                     // Find the ray with the greatest distance from the axis
-                    // Should we use the average distance instead?
                     float radius = 1.0f;
                     for (const auto& idx : cluster) {
                         const Eigen::Vector3f& ray(rays.col(idx));
-                        if (axis.dot(ray) < radius) {
-                            radius = axis.dot(ray);
-                        }
+                        // arccos of the dot product gives the angle between the rays
+                        // the smaller arccos is, the bigger the angle - if between 0 and pi
+                        radius = axis.dot(ray) < radius ? axis.dot(ray) : radius;
```

Comment:
Ideally, this part of the code would use the cone radius once a cone was fit to the edges.

Also, this comment describes what is happening but not why. 
A good comment will explain the things you can't get by reading the code.
Making a comment that just describes what code is doing is bad because if the code changes in the future and not the comment it'll get very confusing and reading the code can give you the information you need.

In this case, we are trying to find the largest ball radius that is possible based on the edge points we have seen from the central axis.
The idea is that the cone that contains all the ball points from the centre of the ball is most likely to be the true radius of the ball since you don't have any points that are outside of that cone.
The way the visual mesh works is that you'll have points on the ball but not outside the ball, so the largest distance between the centre of the ball and a point that is ball is hopefully as close to the edge as possible.

If you did want to improve this part as well, you can look at the neighbours in the visual mesh for each of these points (which are not ball) and you know those are beyond the edge, so between the points that are ball and not ball you have a bound of where the edge of the ball lives

Reply:
Once we have a benchmarking test set up in Webots, I might have a go at improving this given what you've said. Then I can see if any changes really improves it. 

I get what this means more or less now, but how would you describe this radius? Since it's not the true radius, since it comes from the unit vectors. Is it the approximate radius of the ball on the image itself? Or is there a better way to describe it?